src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -493,6 +495,33 @@ public static string GetMSBuildFilePath(PackageSpec project, RestoreRequest requ
 
                 props.AddRange(GenerateGroupsWithConditions(buildPropsGroup, isMultiTargeting, frameworkConditions));
 
+                // Create an empty PropertyGroup for package properties
+                var packagePathsPropertyGroup = MSBuildRestoreItemGroup.Create("PropertyGroup", Enumerable.Empty<XElement>(), 1000, isMultiTargeting ? frameworkConditions : Enumerable.Empty<string>());
+
+                var projectGraph = targetGraph.Graphs.FirstOrDefault();
+
+                // Distinct union of tool packages and packages with GeneratePathProperty=true
+                var packageIdsToCreatePropertiesFor = packagesWithTools.Union(projectGraph.Item.Data.Dependencies.Where(i => i.GeneratePathProperty).Select(i => i.Name)).Distinct(StringComparer.OrdinalIgnoreCase);
+
+                // Find the packages with matching IDs in the list of sorted packages, filtering out ones that there was no match for or that don't exist
+                var packagePathProperties = packageIdsToCreatePropertiesFor
+                    // Cast sortedPackages as nullable
+                    .Select(packageId => sortedPackages.Cast<KeyValuePair<LockFileTargetLibrary, Lazy<LocalPackageSourceInfo>>?>().FirstOrDefault(pkg => pkg != null && pkg.Value.Key.Name.Equals(packageId, StringComparison.OrdinalIgnoreCase)))
+                    // Check if there was a match found and the package exists on disk
+                    .Where(pkg => pkg != null && pkg.Value.Value.Exists())
+                    // Select just the value of the nullable KeyValuePair<LockFileTargetLibrary, Lazy<LocalPackageSourceInfo>>
+                    .Select(pkg => pkg.Value)
+                    // Get the property
+                    .Select(i => GeneratePackagePathProperty(i.Key.Name, i.Value.Value.Package.ExpandedPath));
```

Comment:
You can make GeneratePackagePathProperty accept a localpackage as a paremeter to make it cleaner. 

The Package has an Id afaik.